### PR TITLE
Use logical positioning instead of physical for bidirectional text support

### PR DESCRIPTION
**Pull request by AndyScherzinger**
* Successor of #42471
* For: #31420 
  * Backend part moved to #47349
---

## Summary
This PR is the first step towards adding bidirectional text support to the Nextcloud ecosystem. It affects the entire project; accordingly, each app should have subsequent PRs.

The PR is 90+ percent complete. I decided to make the PR to
- Get early feedback to ensure it is on the right path.
- Discuss some technical challenges.

I had challenges in building the project. Generating `css` files and the `dist` directory during development is something I am uncomfortable with.

Because of supporting Samsung Internet, I needed to take some hacky approach.

**Note:**
We must update many components in [NextCloud/vue](https://github.com/nextcloud-libraries/nextcloud-vue/). The most important one was the `header-menu` vue component which is [already updated](https://github.com/nextcloud-libraries/nextcloud-vue/pull/4965). Others will come soon. First, I need to ensure this PR is getting prepared for merge.

## Screenshots
| Before | After |
|---|---|
|image|image|
|image|image|

## TODO

- [x] Decide about some classes like `.right` or `.menu-left`.
- [ ] Decide about `core/css/tooltip.scss` and its right|left classes
- [ ] Decide about https://github.com/nextcloud/server/blob/6b71288a30d6acbaaeb2f933a44883f3f51d7260/core/src/views/Profile.vue inline styles (for background-position.
- [ ] Add test to ensure we have proper `dir` attribute on the body. (help needed)

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

**Comment by skjnldsv** [comment]
Files potentially missing from previous PR:

- `apps/files_reminders/src/components/setcustomremindermodal.vue`
- `apps/files_sharing/src/components/sharingentryquickshareselect.vue`
- `apps/settings/src/components/applist/appitem.vue`
- `apps/settings/src/components/personalinfo/emailsection/email.vue`
- `apps/settings/src/views/apps.vue`
- `apps/updatenotification/src/components/updatenotification.vue`
- `apps/user_status/src/userstatus.vue`
- `core/src/components/usermenu/usermenuentry.vue`
- `core/src/views/usermenu.vue`

File: lib/private/L10N/Factory.php [hunk 1]
```diff
@@ -364,6 +370,15 @@ public function languageExists($app, $lang) {
 		return in_array($lang, $languages);
 	}
 
+	public function getLanguageDirectionFromLocale($locale): string
```
Comment (susnux) [inline lib/private/L10N/Factory.php:373]:
```suggestion
	/**
	 * @return 'ltr'|'rtl'
	 */
	public function getLanguageDirectionFromLocale($locale): string
```

File: lib/public/L10N/IFactory.php [hunk 2]
```diff
@@ -103,6 +103,14 @@ public function languageExists($app, $lang);
 	 */
 	public function localeExists($locale);
 
+	/**
+	 * Return the current language direction from locale
+	 *
+	 * @param string $locale
+	 * @return string
```
Comment (susnux) [inline lib/public/L10N/IFactory.php:110]:
```suggestion
	 * @return string
	 * @since 31.0.0
```

**Comment by nickvergessen** [comment]
Backend part moved to https://github.com/nextcloud/server/pull/47349

File: core/css/styles.scss [hunk 3]
```diff
@@ -485,11 +492,10 @@ code {
...
 
-			&.ui-icon-circle-triangle-e {
-				background: url("../img/actions/arrow-right.svg") center center no-repeat;
-			}
+			&.ui-icon-circle-triangle-e,
 			&.ui-icon-circle-triangle-w {
-				background: url("../img/actions/arrow-left.svg") center center no-repeat;
+				background-position: center center;
+				background-repeat: no-repeat;
 			}
```
Comment (susnux) [inline core/css/styles.scss:499]: This seems to be very unrelated

File: core/src/components/Profile/PrimaryActionButton.vue [hunk 4]
```diff
@@ -61,4 +61,37 @@ export default defineComponent({
...
+				filter: invert(1);
+			}
+		}
+
+		&:hover,
+		&:focus,
+		&:active {
+			background-color: var(--color-primary-element-light);
+		}
+	}
```
Comment (susnux) [inline core/src/components/Profile/PrimaryActionButton.vue:96]: This is unrelated and should be removed

File: core/src/views/Profile.vue [hunk 5]
```diff
@@ -311,7 +311,9 @@ $content-max-width: 640px;
 		align-self: flex-start;
 		padding-top: 20px;
 		min-width: 220px;
-		margin: -150px 20px 0 0;
+		margin: 0;
+		margin-top: -150px;
+		margin-inline-end: 20px;
```
Comment (susnux) [inline core/src/views/Profile.vue:316]: Maybe better:
```suggestion
		margin-block: -150px 0;
		margin-inline: 0 20px;
```

File: core/src/components/Profile/PrimaryActionButton.vue (hunk 4 above)
Comment (susnux) [reply core/src/components/Profile/PrimaryActionButton.vue:96]:
```suggestion
```

File: apps/settings/src/components/AdminTwoFactor.vue [hunk 6]
```diff
@@ -175,8 +175,8 @@ export default {
 	.two-factor-loading {
 		display: inline-block;
 		vertical-align: sub;
-		margin-left: -2px;
-		margin-right: 1px;
+		margin-inline-start: -2px;
+		margin-inline-end: 1px;
```
Comment (susnux) [inline apps/settings/src/components/AdminTwoFactor.vue:179]:
```suggestion
		margin-inline: -2px 1px;
```

File: apps/settings/src/components/AppList/AppItem.vue [hunk 7]
```diff
@@ -257,8 +257,8 @@ export default {
 		.app-name--link::after {
 			content: '';
 			position: absolute;
-			left: 0;
-			right: 0;
+			inset-inline-start: 0;
+			inset-inline-end: 0;
```
Comment (susnux) [inline apps/settings/src/components/AppList/AppItem.vue:261]:
```suggestion
			inset-inline: 0;
```

File: apps/settings/src/components/Markdown.vue [hunk 8]
```diff
@@ -177,11 +177,11 @@ export default {
...
 	blockquote {
-		padding-left: 1em;
-		border-left: 4px solid var(--color-primary-element);
+		padding-inline-start: 1em;
+		border-inline-start: 4px solid var(--color-primary-element);
 		color: var(--color-text-maxcontrast);
-		margin-left: 0;
-		margin-right: 0;
+		margin-inline-start: 0;
+		margin-inline-end: 0;
```
Comment (susnux) [inline apps/settings/src/components/Markdown.vue:184]:
```suggestion
		margin-inline: 0;
```

File: apps/settings/src/components/WebAuthn/AddDevice.vue [hunk 9]
```diff
@@ -179,8 +179,8 @@ export default {
 .webauthn-loading {
 	display: inline-block;
 	vertical-align: sub;
-	margin-left: 2px;
-	margin-right: 2px;
+	margin-inline-start: 2px;
+	margin-inline-end: 2px;
```
Comment (susnux) [inline apps/settings/src/components/WebAuthn/AddDevice.vue:183]:
```suggestion
	margin-inline: 2px;
```

File: apps/workflowengine/src/styles/operation.scss [hunk 10]
```diff
@@ -7,10 +7,10 @@
 	flex-wrap: wrap;
 	flex-direction: column;
 	flex-grow: 1;
-	margin-left: -1px;
+	margin-inline-start: -1px;
```
Comment (susnux) [inline apps/workflowengine/src/styles/operation.scss:10]:
```suggestion
```

File: apps/workflowengine/src/styles/operation.scss [hunk 11]
```diff
@@ -7,10 +7,10 @@
...
 	padding: 10px;
 	border-radius: var(--border-radius-large);
-	margin-right: 20px;
+	margin-inline-end: 20px;
```
Comment (susnux) [inline apps/workflowengine/src/styles/operation.scss:13]:
```suggestion
	margin-inline: -1px 20px;
```

File: core/css/apps.scss [hunk 12]
```diff
@@ -549,8 +549,8 @@ kbd {
 	 * Editable entries
 	 */
 	.app-navigation-entry-edit {
-		padding-left: 5px;
-		padding-right: 5px;
+		padding-inline-start: 5px;
+		padding-inline-end: 5px;
```
Comment (susnux) [inline core/css/apps.scss:553]:
```suggestion
		padding-inline: 5px;
```

File: core/css/apps.scss [hunk 13]
```diff
@@ -1390,13 +1445,13 @@ $outter-margin: math.div($popoveritem-height - $popovericon-size, 2);
...
 			display: block;
-			padding-left: 50px;
+			padding-inline-start: 50px;
 			white-space: nowrap;
 			overflow: hidden;
 			text-overflow: ellipsis;
 			order: 1;
 			flex: 1 1 0px;
-			padding-right: 10px;
+			padding-inline-end: 10px;
```
Comment (susnux) [inline core/css/apps.scss:1454]:
```suggestion
```

File: core/css/apps.scss [hunk 14]
```diff
@@ -1390,13 +1445,13 @@ $outter-margin: math.div($popoveritem-height - $popovericon-size, 2);
 		.app-content-list-item-line-one,
 		.app-content-list-item-line-two {
 			display: block;
-			padding-left: 50px;
+			padding-inline-start: 50px;
```
Comment (susnux) [inline core/css/apps.scss:1448]:
```suggestion
			padding-inline: 50px 10px;
```

File: core/css/guest.scss [hunk 15]
```diff
@@ -255,12 +255,12 @@ input[type='email'] {
 	cursor: text;
 	font-family: inherit;
 	font-weight: normal;
-	margin-left: 0;
-	margin-right: 0;
+	margin-inline-start: 0;
+	margin-inline-end: 0;
```
Comment (susnux) [inline core/css/guest.scss:259]:
```suggestion
	margin-inline: 0;
```

File: core/css/guest.scss [hunk 16]
```diff
@@ -582,21 +586,26 @@ p.info {
...
-		text-align: left;
-		margin-left: 25px;
-		margin-right: 25px;
-	}
-
-	img.float-spinner {
-		float: left;
+		text-align: start;
+		margin-inline-start: 25px;
+		margin-inline-end: 25px;
```
Comment (susnux) [inline core/css/guest.scss:591]:
```suggestion
		margin-inline: 25px;
```

File: core/css/systemtags.scss [hunk 17]
```diff
@@ -8,8 +8,8 @@
 	.select2-result-label {
 		.checkmark {
 			visibility: hidden;
-			margin-left: -5px;
-			margin-right: 5px;
+			margin-inline-start: -5px;
+			margin-inline-end: 5px;
```
Comment (susnux) [inline core/css/systemtags.scss:12]:
```suggestion
			margin-inline: -5px 5px;
```

File: core/src/views/LegacyUnifiedSearch.vue [hunk 18]
```diff
@@ -745,7 +745,9 @@ $input-padding: 10px;
 	}
 
 	&__filters {
-		margin: $margin 0 $margin math.div($margin, 2);
+		margin-block: $margin;
+		margin-inline-end: 0;
+		margin-inline-start: math.div($margin, 2);
```
Comment (susnux) [inline core/src/views/LegacyUnifiedSearch.vue:750]:
```suggestion
		margin-inline: math.div($margin, 2) 0;
```

File: core/src/views/LegacyUnifiedSearch.vue [hunk 19]
```diff
@@ -760,8 +762,8 @@ $input-padding: 10px;
 
 		// Loading spinner
 		&::after {
-			right: $input-padding;
-			left: auto;
+		inset-inline-end: $input-padding;
+		inset-inline-start: auto;
```
Comment (susnux) [inline core/src/views/LegacyUnifiedSearch.vue:766]:
```suggestion
		inset-inline-start: auto $input-padding;
```

**Review by susnux — COMMENTED** [review]
Test seems to work fine, but some comments.

**Review by susnux — APPROVED** [review]
Tested it, could not find any regressions with LTR language.
So I think we should be able to merge even before the backend PR which allows RTL layout.

**Comment by susnux** [comment]
> Ready for review ?

Yes I think so

**Review by skjnldsv — APPROVED** [review]
